File: ios/agora_rtc_engine/Sources/agora_rtc_engine/AgoraRTE.m
#import "AgoraRTE.h"
#import "AgoraRTECommon.h"

@interface AgoraRTE ()
@property (nonatomic, strong, readwrite) AgoraRte *rteInstance;
@end

@implementation AgoraRTE

- (BOOL)getFromBridge:(NSError **)error {
    AgoraRteError *rteError = [[AgoraRteError alloc] init];
    self.rteInstance = [AgoraRte getFromBridge:rteError];
    
    if (!self.rteInstance || rteError.code != AgoraRteOk) {
        if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
        return NO;
    }
    
    return YES;
}

- (BOOL)createWithConfig:(NSDictionary *)config error:(NSError **)error {
    self.rteInstance = [[AgoraRte alloc] initWithInitialConfig:nil];
    
    if (!self.rteInstance) {
        if (error) *error = RTE_NSERROR(-1, @"Failed to create RTE instance");
        return NO;
    }
    
    if (config && config.count > 0) {
        return [self setConfigs:config error:error];
    }
    
    return YES;
}

- (BOOL)initMediaEngine:(void (^)(NSError *error))completion error:(NSError **)error {
    if (!self.rteInstance) {
        if (error) *error = RTE_NSERROR(-1, @"RTE instance not created");
        return NO;
    }
    
    AgoraRteError *rteError = [[AgoraRteError alloc] init];
    BOOL success = [self.rteInstance initMediaEngine:^(AgoraRteError *err) {
        AgoraRteError *rteErrorCopy = [[AgoraRteError alloc] init];
        [rteErrorCopy setErrorWithCode:err.code message:err.message];
        NSError *nsError = nil;
        if (rteErrorCopy && rteErrorCopy.code != AgoraRteOk) {
            nsError = RTE_NSERROR_FROM_RTE_ERROR(rteErrorCopy);
        }
        if (completion) completion(nsError);
    } error:rteError];
    
    if (!success || rteError.code != AgoraRteOk) {
        if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
        return NO;
    }
    return YES;
}

- (BOOL)destroy:(NSError **)error {
    if (!self.rteInstance) return YES;
    
    AgoraRteError *rteError = [[AgoraRteError alloc] init];
    BOOL success = [self.rteInstance destroy:rteError];
    
    if (!success || rteError.code != AgoraRteOk) {
        if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
        return NO;
    }
    
    self.rteInstance = nil;
    return YES;
}
- (BOOL)setConfigs:(NSDictionary *)config error:(NSError **)error {
    CHECK_RTE_INSTANCE(self.rteInstance, error);
    
    
    AgoraRteConfig *rteConfig = [[AgoraRteConfig alloc] init];
    AgoraRteError *rteError = [[AgoraRteError alloc] init];
    
    if (config[@"appId"] && config[@"appId"] != [NSNull null]) {
        [rteConfig setAppId:config[@"appId"] error:rteError];
        if (rteError.code != AgoraRteOk) {
            if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
            return NO;
        }
    }
    if (config[@"logFolder"] && config[@"logFolder"] != [NSNull null]) {
        [rteConfig setLogFolder:config[@"logFolder"] error:rteError];
        if (rteError.code != AgoraRteOk) {
            if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
            return NO;
        }
    }
    if (config[@"logFileSize"] && config[@"logFileSize"] != [NSNull null]) {
        [rteConfig setLogFileSize:[config[@"logFileSize"] unsignedLongValue] error:rteError];
        if (rteError.code != AgoraRteOk) {
            if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
            return NO;
        }
    }
    if (config[@"areaCode"] && config[@"areaCode"] != [NSNull null]) {
        [rteConfig setAreaCode:[config[@"areaCode"] intValue] error:rteError];
        if (rteError.code != AgoraRteOk) {
            if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
            return NO;
        }
    }
    if (config[@"cloudProxy"] && config[@"cloudProxy"] != [NSNull null]) {
        [rteConfig setCloudProxy:config[@"cloudProxy"] error:rteError];
        if (rteError.code != AgoraRteOk) {
            if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
            return NO;
        }
    }
    if (config[@"jsonParameter"] && config[@"jsonParameter"] != [NSNull null]) {
        [rteConfig setJsonParameter:config[@"jsonParameter"] error:rteError];
        if (rteError.code != AgoraRteOk) {
            if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
            return NO;
        }
    }
    if (config[@"useStringUid"] && config[@"useStringUid"] != [NSNull null]) {
        [rteConfig setUseStringUid:[config[@"useStringUid"] boolValue] error:rteError];
        if (rteError.code != AgoraRteOk) {
            if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
            return NO;
        }
    }
    AgoraRteError *setError = [[AgoraRteError alloc] init];
    BOOL success = [self.rteInstance setConfigs:rteConfig error:setError];
    
    if (!success || setError.code != AgoraRteOk) {
        if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(setError);
        return NO;
    }
    return YES;
}

- (NSDictionary *)getConfigs:(NSError **)error {
    CHECK_RTE_INSTANCE_NIL(self.rteInstance, error);
    
    AgoraRteConfig *config = [[AgoraRteConfig alloc] init];
    AgoraRteError *rteError = [[AgoraRteError alloc] init];
    
    BOOL success = [self.rteInstance getConfigs:config error:rteError];
    if (!success || rteError.code != AgoraRteOk) {
        if (error) *error = RTE_NSERROR_FROM_RTE_ERROR(rteError);
        return nil;
    }
    
    AgoraRteError *getError = [[AgoraRteError alloc] init];
    BOOL useStringUidValue = [config useStringUid:getError];
    return @{
        @"appId": [config appId:getError] ?: @"",
        @"logFolder": [config logFolder:getError] ?: @"",
        @"logFileSize": @([config logFileSize:getError]),
        @"areaCode": @([config areaCode:getError]),
        @"cloudProxy": [config cloudProxy:getError] ?: @"",
        @"jsonParameter": [config jsonParameter:getError] ?: @"",
        @"useStringUid": [NSNumber numberWithBool:useStringUidValue]
    };
}

@end

File: android/src/main/java/io/agora/agora_rtc_ng/AgoraRTE.java
package io.agora.agora_rtc_ng;

import io.agora.rte.Rte;
import io.agora.rte.InitialConfig;
import io.agora.rte.Config;
import io.agora.rte.callback.AsyncCallback;
import io.agora.rte.exception.RteException;

import java.util.HashMap;
import java.util.Map;

/**
 * RTE main class, manages RTE lifecycle
 */
public class AgoraRTE {
    private Rte rteInstance;
    /** Cache for uint32 values that overflow int (SDK uses int). */
    private Long lastSetLogFileSize;
    private Long lastSetAreaCode;

    public AgoraRTE() {
        this.rteInstance = null;
    }

    /**
     * Create RTE instance from Bridge
     */
    public boolean getFromBridge() {
        try {
            rteInstance = Rte.getFromBridge();
            return true;
        } catch (RteException e) {
            rteInstance = null;
            return false;
        }
    }

    /**
     * Create RTE instance with configuration
     */
    public boolean createWithConfig(Map<String, Object> configMap) {
        if (rteInstance != null) {
            return false;
        }
        InitialConfig initialConfig = new InitialConfig();
        rteInstance = new Rte(initialConfig);
        
        if (configMap != null && !configMap.isEmpty()) {
            if (configMap.containsKey("appId")) {
                Object appIdObj = configMap.get("appId");
                if (appIdObj == null || (appIdObj instanceof String && ((String) appIdObj).isEmpty())) {
                    return false;
                }
            }
            return setConfigs(configMap);
        }
        return true;
    }

    /**
     * Initialize media engine
     */
    public boolean initMediaEngine(AsyncCallback completion) {
        if (rteInstance == null) {
            return false;
        }
        try {
            rteInstance.initMediaEngine(completion);
            return true;
        } catch (RteException e) {
            return false;
        }
    }

    /**
     * Batch set configurations
     */
    public boolean setConfigs(Map<String, Object> configMap) {
        if (rteInstance == null) {
            return false;
        }
        try {
            Config rteConfig = new Config();
            
            // Only update properties provided in the map.
            if (configMap.containsKey("appId") && configMap.get("appId") != null) {
                rteConfig.setAppId((String) configMap.get("appId"));
            }
            if (configMap.containsKey("logFolder") && configMap.get("logFolder") != null) {
                rteConfig.setLogFolder((String) configMap.get("logFolder"));
            }
            if (configMap.containsKey("logFileSize") && configMap.get("logFileSize") != null) {
                long v = parseLong(configMap.get("logFileSize"));
                lastSetLogFileSize = v;
                rteConfig.setLogFileSize((int) v);
            }
            if (configMap.containsKey("areaCode") && configMap.get("areaCode") != null) {
                long v = parseLong(configMap.get("areaCode"));
                lastSetAreaCode = v;
                rteConfig.setAreaCode((int) v);
            }
            if (configMap.containsKey("cloudProxy") && configMap.get("cloudProxy") != null) {
                rteConfig.setCloudProxy((String) configMap.get("cloudProxy"));
            }
            if (configMap.containsKey("jsonParameter") && configMap.get("jsonParameter") != null) {
                rteConfig.setJsonParameter((String) configMap.get("jsonParameter"));
            }
            if (configMap.containsKey("useStringUid") && configMap.get("useStringUid") != null) {
                Object useStringUidObj = configMap.get("useStringUid");
                if (useStringUidObj instanceof Boolean) {
                    rteConfig.setUseStringUid((Boolean) useStringUidObj);
                }
            }
            
            rteInstance.setConfigs(rteConfig);
            return true;
        } catch (RteException e) {
            return false;
        }
    }

    /**
     * Get all configurations
     */
    public Map<String, Object> getConfigs() {
        if (rteInstance == null) {
            return null;
        }
        try {
            Config config = new Config();
            rteInstance.getConfigs(config);
            
            Map<String, Object> map = new HashMap<>();
            map.put("appId", config.getAppId() != null ? config.getAppId() : "");
            map.put("logFolder", config.getLogFolder() != null ? config.getLogFolder() : "");
            map.put("logFileSize", lastSetLogFileSize != null ? lastSetLogFileSize : (long) config.getLogFileSize());
            map.put("areaCode", lastSetAreaCode != null ? lastSetAreaCode : (long) config.getAreaCode());
            map.put("cloudProxy", config.getCloudProxy() != null ? config.getCloudProxy() : "");
            map.put("jsonParameter", config.getJsonParameter() != null ? config.getJsonParameter() : "");
            map.put("useStringUid", config.getUseStringUid());
            return map;
        } catch (RteException e) {
            return null;
        }
    }

    /**
     * Destroy RTE instance
     */
    public boolean destroy() {
        if (rteInstance == null) {
            return true;
        }
        try {
            rteInstance.destroy();
            rteInstance = null;
            return true;
        } catch (RteException e) {
            return false;
        }
    }

    public Rte getRteInstance() {
        return rteInstance;
    }

    private int parseInt(Object obj) {
        if (obj == null) {
            return 0;
        }
        if (obj instanceof Number) {
            return ((Number) obj).intValue();
        }
        if (obj instanceof String) {
            try {
                return Integer.parseInt((String) obj);
            } catch (Exception e) {
                return 0;
            }
        }
        return 0;
    }

    private long parseLong(Object obj) {
        if (obj == null) {
            return 0L;
        }
        if (obj instanceof Number) {
            return ((Number) obj).longValue();
        }
        if (obj instanceof String) {
            try {
                return Long.parseLong((String) obj);
            } catch (Exception e) {
                return 0L;
            }
        }
        return 0L;
    }
}
